lib/msal-browser/test/interaction_client/RedirectClient.spec.ts

```diff
@@ -1026,6 +1026,60 @@ describe("RedirectClient", () => {
+                "iss": "https://sts.windows.net/fa15d692-e9c7-4460-a743-29f2956fd429/",
+                "exp": 1536279024,
+                "name": "abeli",
+                "nonce": "123523",
+                "oid": "05833b6b-aa1d-42d4-9ec0-1b2bb9194438",
+                "sub": "5_J9rSss8-jvt_Icu6ueRNL8xXb8LF4Fsg_KooC2RJQ",
+                "tid": "fa15d692-e9c7-4460-a743-29f2956fd429",
+                "ver": "1.0",
+                "preferred_username": "AbeLincoln@contoso.com"
+            };
+            sinon.stub(AuthToken, "extractTokenClaims").returns(idTokenClaims);
+            const browserCrypto = new CryptoOps();
+            const testLogger = new Logger(loggerOptions);
+            const browserStorage: BrowserCacheManager = new BrowserCacheManager(TEST_CONFIG.MSAL_CLIENT_ID, cacheConfig, browserCrypto, testLogger);
+            browserStorage.setTemporaryCache(PersistentCacheKeys.ADAL_ID_TOKEN, TEST_TOKENS.IDTOKEN_V1);
+            const loginUrlSpy = sinon.spy(AuthorizationCodeClient.prototype, "getAuthCodeUrl");
+            sinon.stub(CryptoOps.prototype, "generatePkceCodes").resolves({
+                challenge: TEST_CONFIG.TEST_CHALLENGE,
+                verifier: TEST_CONFIG.TEST_VERIFIER
+            });
+            sinon.stub(CryptoOps.prototype, "createNewGuid").returns(RANDOM_TEST_GUID);
+            sinon.stub(NavigationClient.prototype, "navigateExternal").callsFake((urlNavigate: string, options: NavigationOptions): Promise<boolean> => {
+                expect(options.noHistory).toBeFalsy();
+                expect(urlNavigate).not.toBe("");
+                return Promise.resolve(true);
+            });
+            const emptyRequest: CommonAuthorizationUrlRequest = {
+                redirectUri: TEST_URIS.TEST_REDIR_URI,
+                scopes: [],
+                state: TEST_STATE_VALUES.USER_STATE,
+                authority: TEST_CONFIG.validAuthority,
+                correlationId: TEST_CONFIG.CORRELATION_ID,
+                responseMode: TEST_CONFIG.RESPONSE_MODE as ResponseMode,
+                nonce: "",
+                authenticationScheme: TEST_CONFIG.TOKEN_TYPE_BEARER as AuthenticationScheme
+            };
+            await redirectClient.acquireToken(emptyRequest);
+            const validatedRequest: CommonAuthorizationUrlRequest = {
+                ...emptyRequest,
+                scopes: [],
+                loginHint: idTokenClaims.preferred_username,
+                state: TEST_STATE_VALUES.TEST_STATE_REDIRECT,
+                correlationId: RANDOM_TEST_GUID,
+                nonce: RANDOM_TEST_GUID,
+                authority: `${Constants.DEFAULT_AUTHORITY}`,
+                responseMode: ResponseMode.FRAGMENT,
+                codeChallenge: TEST_CONFIG.TEST_CHALLENGE,
+                codeChallengeMethod: Constants.S256_CODE_CHALLENGE_METHOD
+            };
+            expect(loginUrlSpy.calledWith(validatedRequest)).toBeTruthy();
+        });
+
```

Comment:
```suggestion

        it("Uses adal token from cache if it is present and sets preferred_name as the login hint when upn is also populated.", async () => {
            const idTokenClaims: TokenClaims = {
                "iss": "https://sts.windows.net/fa15d692-e9c7-4460-a743-29f2956fd429/",
                "exp": 1536279024,
                "name": "abeli",
                "nonce": "123523",
                "oid": "05833b6b-aa1d-42d4-9ec0-1b2bb9194438",
                "sub": "5_J9rSss8-jvt_Icu6ueRNL8xXb8LF4Fsg_KooC2RJQ",
                "tid": "fa15d692-e9c7-4460-a743-29f2956fd429",
                "ver": "1.0",
                "upn": "testupn@contoso.com",
                "preferred_username": "AbeLincoln@contoso.com"
            };
            sinon.stub(AuthToken, "extractTokenClaims").returns(idTokenClaims);
            const browserCrypto = new CryptoOps();
            const testLogger = new Logger(loggerOptions);
            const browserStorage: BrowserCacheManager = new BrowserCacheManager(TEST_CONFIG.MSAL_CLIENT_ID, cacheConfig, browserCrypto, testLogger);
            browserStorage.setTemporaryCache(PersistentCacheKeys.ADAL_ID_TOKEN, TEST_TOKENS.IDTOKEN_V1);
            const loginUrlSpy = sinon.spy(AuthorizationCodeClient.prototype, "getAuthCodeUrl");
            sinon.stub(CryptoOps.prototype, "generatePkceCodes").resolves({
                challenge: TEST_CONFIG.TEST_CHALLENGE,
                verifier: TEST_CONFIG.TEST_VERIFIER
            });
            sinon.stub(CryptoOps.prototype, "createNewGuid").returns(RANDOM_TEST_GUID);
            sinon.stub(NavigationClient.prototype, "navigateExternal").callsFake((urlNavigate: string, options: NavigationOptions): Promise<boolean> => {
                expect(options.noHistory).toBeFalsy();
                expect(urlNavigate).not.toBe("");
                return Promise.resolve(true);
            });
            const emptyRequest: CommonAuthorizationUrlRequest = {
                redirectUri: TEST_URIS.TEST_REDIR_URI,
                scopes: [],
                state: TEST_STATE_VALUES.USER_STATE,
                authority: TEST_CONFIG.validAuthority,
                correlationId: TEST_CONFIG.CORRELATION_ID,
                responseMode: TEST_CONFIG.RESPONSE_MODE as ResponseMode,
                nonce: "",
                authenticationScheme: TEST_CONFIG.TOKEN_TYPE_BEARER as AuthenticationScheme
            };
            await redirectClient.acquireToken(emptyRequest);
            const validatedRequest: CommonAuthorizationUrlRequest = {
                ...emptyRequest,
                scopes: [],
                loginHint: idTokenClaims.preferred_username,
                state: TEST_STATE_VALUES.TEST_STATE_REDIRECT,
                correlationId: RANDOM_TEST_GUID,
                nonce: RANDOM_TEST_GUID,
                authority: `${Constants.DEFAULT_AUTHORITY}`,
                responseMode: ResponseMode.FRAGMENT,
                codeChallenge: TEST_CONFIG.TEST_CHALLENGE,
                codeChallengeMet
                hod: Constants.S256_CODE_CHALLENGE_METHOD
            };
            expect(loginUrlSpy.calledWith(validatedRequest)).toBeTruthy();
        });
        
```

Reply:
Example of a second new test to add that includes upn and preferred_username in teh id token.